### PR TITLE
Add GitHub#getInstallation to access endpoints related to the authenticated installation

### LINKED ISSUE #1082: Some "GHAppInstallation" methods need an installation token, not a JWT token
I can explain the oddity on lines 84-87, since I originally wrote them.

https://github.com/hub4j/github-api/blob/7c8a7ff26ebfb31aed3bfdd8f0825ea56142ba4e/src/test/java/org/kohsuke/github/AbstractGHAppInstallationTest.java#L71-L87

The `GitHub` client constructed on line 71 is using the JWT token. Which is appropriate for e.g. retrieving a list of installations for the app and other high-level information about the App.

The tests in `GHAppInstallationTest.java` are using API methods that need a client authenticated as an "Installation", not an "App" though. I.e. a GitHub client needs to be constructed with the "installation token", which is separate and is retrieved using the "JWT token".

Details from the documentation: 
* [Authenticating as a GitHub App](https://docs.github.com/en/developers/apps/authenticating-with-github-apps#authenticating-as-a-github-app)
* [Authenticating as an installation](https://docs.github.com/en/developers/apps/authenticating-with-github-apps#authenticating-as-an-installation)

So the commented out code did exactly that - used the client with the "JWT token" to create an "installation token" and inserted it back on the `GHAppInstallation` object, so further API calls would correctly use the "installation token".

I am pretty sure the live tests will fail now with that code commented out.

And I remember not finding a really nice way (without using `setRoot()`) of exchanging those tokens on the `GHAppInstallation` object, so it is probably an improvement opportunity.

### PR DESCRIPTION
**Pull request by yrodiere**
# Description

Fixes #1082.

Relevant documentation: https://docs.github.com/en/rest/apps/installations#list-repositories-accessible-to-the-app-installation
As you can see, this endpoint can only be reached with an installation token; but `GHAppInstallation` can only be obtained through `GHApp` with a JWT, so its method `listRepositories` just cannot work, except with hacks (`setRoot`, ...). 
That's the problem we're trying to solve.

This fix just adds a method to `GitHub` to obtain an object representing the current installation, which exposes a `listRepositories` method.
Original idea by @dwnusbaum (https://github.com/hub4j/github-api/issues/1082#issuecomment-1197417570).

# Before submitting a PR:

- [X] Changes must not break binary backwards compatibility. If you are unclear on how to make the change you think is needed while maintaining backward compatibility, [CONTRIBUTING.md](CONTRIBUTING.md) for details.
- [X] Add JavaDocs and other comments as appropriate. Consider including links in comments to relevant documentation on https://docs.github.com/en/rest .  
- [X] Add tests that cover any added or changed code. This generally requires capturing snapshot test data. See [CONTRIBUTING.md](CONTRIBUTING.md) for details.
- [X] Run `mvn -D enable-ci clean install site` locally. If this command doesn't succeed, your change will not pass CI.
- [X] Push your changes to a branch other than `main`. You will create your PR from that branch.

# When creating a PR: 

- [X] Fill in the "Description" above with clear summary of the changes. This includes:
  - [X] If this PR fixes one or more issues, include "Fixes #<issue number>" lines for each issue. 
  - [X] Provide links to relevant documentation on https://docs.github.com/en/rest where possible.
- [X] All lines of new code should be covered by tests as reported by code coverage. Any lines that are not covered must have PR comments explaining why they cannot be covered. For example, "Reaching this particular exception is hard and is not a particular common scenario."
- [X] Enable "Allow edits from maintainers".

### CODE REVIEW

**Comment by gsmet** [comment]
Hey @bitwiseman ,

This one is kinda in the way of our work on Quarkus so we would appreciate a lot if you could review and cook a release once everything is fine.

Thanks!

**Review by dwnusbaum — COMMENTED** [review]
Thanks for working on this! It looks good to me in general. I think that adding the `GHAuthenticatedAppInstallation` class is a good idea to avoid polluting the `GitHub` class with methods that only work in some cases. Support for `DELETE /installation/token` can easily be added to the class later, and perhaps GitHub will add other similar `/installation` API endpoints that could live in the new class in the future.

(I am not a maintainer.)

File: src/main/java/org/kohsuke/github/GitHub.java
Comment (yrodiere) [reply src/main/java/org/kohsuke/github/GitHub.java:1175]: Perhaps, but I was under the impression that other methods try to follow the naming of endpoints. And in this case we have:

* `/installation/repositories`
* `getInstallation().listRepositories()`

File: src/main/java/org/kohsuke/github/GHAuthenticatedAppInstallation.java
Comment (yrodiere) [reply src/main/java/org/kohsuke/github/GHAuthenticatedAppInstallation.java:14]: > Is it possible with the current codebase to detect whether the current GitHub client is authenticated as an installation when this object is instantiated on the client side?

Judging from how `GitHub#getApp()` fails with an obscure HTTP authentication error when we try to use it with an installation token, I would say there is no such feature yet. That would probably be a nice improvement in another PR that would be applied across the board to all relevant methods, not just this one? I think the hardest part would probably be testing...

Comment (dwnusbaum) [reply src/main/java/org/kohsuke/github/GHAuthenticatedAppInstallation.java:14]: > {"message":"Bad credentials","documentation_url":"https://docs.github.com/rest"}

Well that is true, but the message is a bit opaque 😄

Either way this is something that could easily be improved in the future if users are actually getting confused by it, so there's probably no reason to complicate this PR.

**Comment by yrodiere** [comment]
Hey @bitwiseman , thanks for allowing the CI build. It looks like the build passes, for this PR as well as for my two other PRs (#1524, #1525).

I'll be happy to address any comment you have on those PRs.

If the PR look good to you, do you think you could merge them and release a new version of github-api? As Guillaume mentioned earlier, we would need these patches in a Quarkus-related project :)

**Review by bitwiseman — APPROVED** [review]

**Comment by bitwiseman** [comment]
@yrodiere 
Nice work!

**Comment by bitwiseman** [comment]
Release coming soon.

**Comment by yrodiere** [comment]
Thanks @bitwiseman!

**Comment by bitwiseman** [comment]
@yrodiere 
I'm blocked on publishing a new release by #1528 .  I've spent for more time on this project tonight than I can spare, but I apparently can't produce a new release until that issue is fixed. 😡

**Comment by bitwiseman** [comment]
I was able to release by doing some hacking.  Enjoy!

**Comment by yrodiere** [comment]
Great to see that @bitwiseman , thanks! We're already using it in production :)
I opened #1531 to address the javadoc issue.

**Pull request merged**
